include/openthread/thread.h

```diff
@@ -380,30 +380,52 @@ otError otThreadSetLinkMode(otInstance *aInstance, otLinkModeConfig aConfig);
 /**
  * Get the thrNetworkKey.
  *
+ * This function will return an union of key reference and literal key. These feilds
+ * will be populated based on the Crypto module being used. For PSA Crypto, a reference
+ * to the thrNetworkKey is returned and not the literal key.
```

Comment:
When you use this function, you would need to guard it with some compile-time conditions like:

```
#if OPENTHREAD_TFM
key_ref = keyMaterial.keyRef;
...
#else
memcpy(key_val, keyMaterial.key, KEY_SIZE);
...
#endif
```

And you would need to make sure this code is compiled with the same preprocessor defines that were used during build of OpenThread modules. If you use OpenThread from pre-compiled static libraries, it is easy to introduce build time bug.

I think the API would be cleaner if `otNetworkKey` structure contained a field indicating if the union holds key value or reference. This way, API user would not need to guess what value is returned (if OT library was compiled with or without PSA Crypto support?), but check it run-time like:

```
if (keyMaterial.type == OPENTHREAD_KEYMATERIAL_TYPE_PSA) {
 // Use PSA key reference
} else {
  // Use key value
}
```

Reply:
We won't need the compile time conditions for otThreadGetNetworkKey. It will return the union irrespective of the module used.`otPlatCryptoGetType` to determine if the system supports key references and can export the key. As an alternative I have also added `otCopyNetworkKey`, which returns the network key as a literal irrespective of what Crypto is used.

I don't think we can add a variable to indicate the type in `otThreadGetNetworkKey ` as this is a basic data structure used to generate TLVs. However there is already such flag available in `NetworkKey` class, we can use this and add a output pointer for key type in `otThreadGetNetworkKey `.

Or I can replace the implementation of `otThreadGetNetworkKey` with `otThreadCopyNetworkKey`. This way user will not have to worry about what Crypto is used internally and he will get the Network key as a literal everytime.

Reply:
> `otPlatCryptoGetType` to determine if the system supports key references

I'm not sure if user of OpenThread should call any of `otPlat...` functions. These are platform function and should be called by OpenThread modules. Not by its user. We could introduce a new function like `otCryptoGetType()` that would be used by the user, but then we would need to document in this function: `If @ref otCryptoGetType() returns X, field A from returned union is valid. If it returns Y, field B from returned union is valid...`.

> we can use this and add a output pointer for key type in otThreadGetNetworkKey

I think an output pointer would be the easiest to use by API user. It should be easier to document which field in union contains valid value based on the value set in the output pointer.

Reply:
> I'm not sure if user of OpenThread should call any of otPlat... functions. These are platform function and should be called by OpenThread modules. Not by its user. 

I 100% agree. Users should not use `otPlat` APIs. Everything they need should be already provided in public OT headers and definitions.  If we want the users to know what type of crypto is being used, we need to add a public OT API for them get the info. 

> if otNetworkKey structure contained a field indicating if the union holds key value or reference.

I agree with @hubertmis  and was considering the same. Adding `mType` would be cleaner model.

Reply:
Here is another way we can approach this:
- We do not define a `union` type in public header (we should use `union` in the core). 
   - Instead we continue to have `otNetworkKey` as before  (an array of bytes represeting the literal key)
    - And define a new `otNetworkKeyRef` which is the key ref (for PSA mode).
- For APIs that get and set the key we provide different APIs, one using `otNetworkKey` and one using `onNetworkKeyRef`.
    - We keep `otThreadSetNetworkKey(otInstance *, const otNetworkKey *)` as is
    - And add `otThreadSetNetworkKeyFromRef(otInstance *, otNetworkKeyRef aKeyRef)`.
    - Similar way for getting the key.
 - The implementation for these API is then provided only when the feature is supported (tied to related `OPENTHREAD_CONFIG_{config}`) and we can document this in the public API description.


The advantages I see with this:
- It is simpler for users of API to use.
- The presence of implementation an API being tied to a `OT_CONFIG` aligns well with how it is done with many other OT APIs today (e.g, `otIp6SetSlaacEnabled`, `otDnsClinet{}`, etc).
   - The immediate benefit of this model is if an API that is not supported is used by app layer code, it causes a build/link error which is better than a run-time check.

Reply:
We may want to offer the use of setting the network key as a literal even with PSA. As this will mimic the real products better. In real products often the network key is either generated or derived from TLVs and in both cases the stack has to deal with literal keys and not key refs. Also asking the user set the network key using key ref would mean the user having to access the platform code to import the key himself, or calling the PSA APIs in application. This would create complications when we move the code to secure world, as now you will have to deal with App accessing PSA APIs as well.

I don't see the app ever needing to get the key refs, especially is the the platform APIs to export the key is not available and the key itself is not exportable. 

So I would propose keeping the getters and setters to work only on literal keys and modify these to abstract the key references where-ever necessary. As such I am inclined towards the first approach of adding `mKeyType` to network key and PSKc.

@hubertmis @abtink any thoughts or objections?

Reply:
> We may want to offer the use of setting the network key as a literal even with PSA.

This can be supported with the model above (i.e., having two APIs for set one with literal key and one with key ref). We will have all API deceleration in the header files, but provide implementation/definition for the ones that are supported (which can both versions).

The main challenge/concern I see with the `union` model representation of keys to the user is that it requires/assumes that user would use the API correctly and/or needs to perform run-time checks

With the `union` model, an incorrect use of API (e.g., using the keyref or literal key mode when it is not supported) will still compile/build ok and we can get a run-time error.

With having different APIs  (one using literal key and one with key ref), an incorrect use is detected during build time (linker failure saying there is no implementation for an incorrectly used API). 

----------

> So I would propose keeping the getters and setters to work only on literal keys and modify these to abstract the key references where-ever necessary.

I am fine with having public API only deal with literal keys in this PR. Then I would say we should keep the API as they are in `main` and remove the `union` defiition and any notion of `key-ref` in the public headers. Basically moving the `union` and key-ref related definition to be core-internal types and definitions.

Reply:
> I am fine with having public API only deal with literal keys in this PR. Then I would say we should keep the API as they are in main and remove the union defiition and any notion of key-ref in the public headers. Basically moving the union and key-ref related definition to be core-internal types and definitions.

I agree the above approach, but if we can't keep the current API signature. Because we are abstracting the getter function to hide the key refs from the user, we cannot return the `otNetworkKey` as before, because `otNetworkKey` is not used in the core anymore. So I would like to change the current getter to pass a buffer to copy the key into and `otGetNetworkKey` to either copy from `otNetworkKey` or export using the key ref.

Nothing to this affect:

```otError otThreadGetNetworkKey(otInstance *aInstance, otNetworkKey *aNetworkKey)
{
    Instance &instance = *static_cast<Instance *>(aInstance);

    return instance.Get<KeyManager>().GetNetworkKey().CopyKey(aNetworkKey->mKeyMaterial.mKey, OT_NETWORK_KEY_SIZE);
}```

(`CopyKey` returns an error if the key is not exportable).
 
Without this, there will be no way for the user to read the network key.

Reply:
@hemanth-silabs, yes. Thanks. This  sounds good to me. We just update `otThreadGetNetworkKey()` to require a param to copy the key into, instead of returning the value. With this, I guess we won't need `otThreadCopyNetworkKey()` as well.

Reply:
Yes, we don't. I do have one more question: Do we consider cli to be part of OT core component? Can we use psa APIs in `cli.cpp`?

Reply:
I am just scoping the other solution you proposed and thinking of different scenarios of how to make it work.

Reply:
> I do have one more question: Do we consider cli to be part of OT core component? Can we use PSA APIs in cli.cpp?

Generally, I would say that `cli` is not part of core. We try to use public OT APIs in CLI and NCP.  We want CLI to be considered as an example of app layer on how to use OT APIs.

(there are few cases over years where some core types are leaked in and used in CLI but we try to remove them).

If we need next layer to use PSA, we can cosndier defining a set of public OT APIs for that.

Reply:
> If we need next layer to use PSA, we can cosndier defining a set of public OT APIs for that.

Some applications (like CLI) could require to export key value, based on key reference retrieved from OpenThread API. I understand, it's not responsibility of OpenThread to allow such key value export. It's responsibility of crypto drivers/platform in given system.

Unfortunately, CLI application does not support platform abstractions and it is hard to handle key value exporting in there.

I don't have good ideas how to keep CLI application simple (without introducing platform abstraction) and to export key value in this application.

Reply:
@hubertmis what are your thoughts on instantiating crypto abstraction similar to what happens with radio driver. This way we can create public APIs to import and export keys to and from PSA. Something along the lines of:

```
otError otThreadExportKey(otInstance *aInstance, otCryptoKeyRef aKeyId uint8_t *aBuffer, size_t aBufferLen, size_t *aKeyLength)
{
    Instance &instance = *static_cast<Instance *>(aInstance);

    return instance.Get<CryptoInstance>().ExportKey(aKeyId, aBuffer, aBufferLen, aKeyLength);
}
```

We can define such interfaces for selected Crypto APIs (or all of them if needed. This way applications can also leverage the PSA functionality without too much hassle of dealing with underlying interface.

Reply:
I don't think such API should be exposed by OpenThread, because it is not OpenThread's responsibility to allow exporting keys.

We can agree to add it, as a workaround for CLI application, but I would add a note to such function like "It is used by OpenThread examples. In real products use your crypto API directly."
Also, we can agree to call `otPlatCrypto...` directly from CLI as another exception, but we are trying to avoid it.

I think it would be best to handle key export inside CLI application. We could call PSA Crypto API directly if given compile-time condition like `OPENTHREAD_CRYPTO_PSA` is met.

I don't like any of those proposals, but we need to settle on one to move forward.

Reply:
> Also, we can agree to call `otPlatCrypto...` directly from CLI as another exception, but we are trying to avoid it.

I think this can work. We already have precedence for using platform APIs in CLI (some of the `otPlatRadio` APIs are directly used/called from CLI and NCP, e.g. `otPlatRadioGetCcaEnergyDetectThreshold()`). 

The main thing we want to avoid is using any of core internal types/definition from CLI and want to limit the interaction to well-defined interfaces, like OT public APIs or OT platform APIs.